src/Collie/Modifiers.idr

```diff
@@ -35,32 +34,52 @@ option : String -> Arguments -> Modifier
 option desc ducer = MkOption $ MkRecord [desc, ducer]
 
 public export
-ParsedModifier : Modifier -> Type
-ParsedModifier (MkFlag   flg) = Bool
-ParsedModifier (MkOption opt) = Carrier (opt.project "arguments").domain
+ParsedModifier : (f, g : Type -> Type) -> Modifier -> Type
+ParsedModifier f g (MkFlag   flg) = f Bool
+ParsedModifier f g (MkOption opt) = g (Carrier (opt.project "arguments").domain)
 
 public export
-ParsedModifiers : (mods : Fields Modifier) -> Type
-ParsedModifiers mods = Record (Maybe . ParsedModifier) mods
+ParsedModifiers : (f, g : Type -> Type) -> (mods : Fields Modifier) -> Type
+ParsedModifiers f g mods = Record (ParsedModifier f g) mods
+
+||| All the flags have a default value. If no value has been set then use
+||| that default instead.
+public export
+defaulting : {mods : Fields Modifier} ->
+  ParsedModifiers Maybe f mods -> ParsedModifiers Prelude.id f mods
+defaulting = map $ \ str, mod, val => case mod of
+  MkFlag   flg => fromMaybe (flg.project "default") val
+  MkOption opt => val
+
+public export
+initNothing : {flds : Fields Modifier} -> ParsedModifiers Maybe Maybe flds
+initNothing = MkRecord $ tabulate $ \ (str, mod) => case mod of
+  MkFlag   flg => Nothing
+  MkOption opt => Nothing
 
 public export
 updateModifier : (name : String) -> {mod : Modifier} ->
-  (new : ParsedModifier mod) ->
-  (old : Maybe (ParsedModifier mod)) -> Error (Maybe $ ParsedModifier mod)
-updateModifier  name                      new Nothing    = pure $ Just new
+  (new : ParsedModifier Prelude.id Prelude.id mod) ->
+  (old : ParsedModifier Maybe Maybe mod) ->
+  Error (ParsedModifier Maybe Maybe mod)
+updateModifier  name {mod = MkFlag   flg} new Nothing    = pure (Just new)
+updateModifier  name {mod = MkOption opt} new Nothing    = pure (Just new)
 {- TODO: currently, overwrite previous flag, but we can do something
 better here: customise the behaviour, or parameterise by a partial
 monoid -}
-updateModifier  name {mod = MkFlag   flg} new (Just old) = pure $ Just new
+updateModifier  name {mod = MkFlag   flg} new (Just old) = pure (Just new)
 updateModifier  name {mod = MkOption opt} new (Just old) with ((opt.project "arguments").domain)
  updateModifier name {mod = MkOption opt} new (Just old) | Some d
    = throwE $ "MkOption \{name} set twice"
  updateModifier name {mod = MkOption opt} new (Just old) | ALot ds
    = let _ = openMagma ds in
-     pure $ Just $ old <+> new
+     pure (Just $ old <+> new)
 
 public export
-(.update) : {mods : Fields Modifier} -> (ps : ParsedModifiers mods) ->
+-- TODO: can we generalise this beyond `Maybe`?
```

Comment:
I think the abstraction here is the structure of a monoid (structure) action: we have a monoid for changes, and those changes act on the final value.

Reply:
(not sure it's worth the effort at the moment, but might be useful later one.)